File: tests/xcalablemp/global-view/gmove/C/gmove_bca_8.c
#define NAMELEN 25
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

extern int chk_int(char name[], int ierr);

void gmove_bca_2a3t_b2_subcomm(){

char name[NAMELEN]="gmove_bca_2a3t_b2_subcomm";
int n=4;
int a[n][n],b[n][n];
#pragma xmp nodes p(8)
#pragma xmp nodes p1(2,2)=p(1:4)
#pragma xmp nodes p2(2,2)=p(5:8)
#pragma xmp template tx(0:n-1,0:n-1,0:n-1)
#pragma xmp template ty(0:n-1,0:n-1,0:n-1)
#pragma xmp distribute tx(block,*,block) onto p1
#pragma xmp distribute ty(*,block,block) onto p2
#pragma xmp align a[i0][*] with tx(i0,*,*)
#pragma xmp align b[*][i2] with ty(*,*,i2)

  int i0,i1,i2,ierr;

#pragma xmp loop (i0) on tx(i0,*,*)
  for(i0=0;i0<n;i0++){
    for(i2=0;i2<n;i2++){
      a[i0][i2]=i0+i2+1;
    }
  }

  for(i1=0;i1<n;i1++){
#pragma xmp loop (i2) on ty(*,*,i2)
    for(i2=0;i2<n;i2++){
      b[i1][i2]=0;
    }
  }

#pragma xmp gmove
  b[:][:]=a[:][:];

  ierr=0;
  for(i1=0;i1<n;i1++){
#pragma xmp loop (i2) on ty(*,*,i2)
    for(i2=0;i2<n;i2++){
      ierr=ierr+abs(b[i1][i2]-i1-i2-1);
    }
  }

#pragma xmp reduction (MAX:ierr)
  chk_int(name, ierr);

}

void gmove_bca_2a3t_b(){

char name[NAMELEN]="gmove_bca_2a3t_b";
int n=8;
int a[n][n],b[n][n];
#pragma xmp nodes p(2,2,2)
#pragma xmp template tx(0:n-1,0:n-1,0:n-1)
#pragma xmp distribute tx(block,block,block) onto p
#pragma xmp align a[i0][i1] with tx(i0,i1,*)
#pragma xmp align b[i1][i2] with tx(*,i1,i2)

  int i0,i1,i2,ierr;

#pragma xmp loop (i0,i1) on tx(i0,i1,*)
  for(i0=0;i0<n;i0++){
    for(i1=0;i1<n;i1++){
      a[i0][i1]=i0+i1+1;
    }
  }

#pragma xmp loop (i1,i2) on tx(*,i1,i2)
  for(i1=0;i1<n;i1++){
    for(i2=0;i2<n;i2++){
      b[i1][i2]=0;
    }
  }

#pragma xmp gmove
  b[:][:]=a[:][:];

  ierr=0;
#pragma xmp loop (i1,i2) on tx(*,i1,i2)
  for(i1=0;i1<n;i1++){
    for(i2=0;i2<n;i2++){
      ierr=ierr+abs(b[i1][i2]-i1-i2-1);
    }
  }

#pragma xmp reduction (MAX:ierr)
  chk_int(name, ierr);

}

void gmove_bca_3a3t_b2(){

char name[NAMELEN]="gmove_bca_3a3t_b2";
int n=4;
int a[n][n][n],b[n][n][n];
#pragma xmp nodes p(2,2,2)
#pragma xmp template tx(0:n-1,0:n-1,0:n-1)
#pragma xmp distribute tx(block,block,block) onto p
#pragma xmp align a[i0][i1][i2] with tx(i0,i1,i2)
#pragma xmp align b[*][i1][i2] with tx(*,i1,i2)

  int i0,i1,i2,ierr;

#pragma xmp loop (i0,i1,i2) on tx(i0,i1,i2)
  for(i0=0;i0<n;i0++){
    for(i1=0;i1<n;i1++){
      for(i2=0;i2<n;i2++){
        a[i0][i1][i2]=i0+i1+i2+1;
      }
    }
  }

#pragma xmp loop (i1,i2) on tx(*,i1,i2)
  for(i0=0;i0<n;i0++){
    for(i1=0;i1<n;i1++){
      for(i2=0;i2<n;i2++){
        b[i0][i1][i2]=0;
      }
    }
  }

#pragma xmp gmove
  b[:][:][:]=a[:][:][:];

  ierr=0;
#pragma xmp loop (i1,i2) on tx(*,i1,i2)
  for(i0=0;i0<n;i0++){
    for(i1=0;i1<n;i1++){
      for(i2=0;i2<n;i2++){
        ierr=ierr+abs(b[i0][i1][i2]-i0-i1-i2-1);
      }
    }
  }

#pragma xmp reduction (MAX:ierr)
  chk_int(name, ierr);

}

void gmove_bca_3a3t_b(){

char name[NAMELEN]="gmove_bca_3a3t_b";
int n=4;
int a[n][n][n],b[n][n][n];
#pragma xmp nodes p(2,2,2)
#pragma xmp template tx(0:n-1,0:n-1,0:n-1)
#pragma xmp distribute tx(block,block,block) onto p
#pragma xmp align a[i0][i1][i2] with tx(i0,i1,i2)
#pragma xmp align b[*][i1][i2] with tx(*,i1,i2)

  int i0,i1,i2,ierr;

#pragma xmp loop (i0,i1,i2) on tx(i0,i1,i2)
  for(i0=0;i0<n;i0++){
    for(i1=0;i1<n;i1++){
      for(i2=0;i2<n;i2++){
        a[i0][i1][i2]=i0+i1+i2+1;
      }
    }
  }

#pragma xmp loop (i1,i2) on tx(*,i1,i2)
  for(i0=0;i0<n;i0++){
    for(i1=0;i1<n;i1++){
      for(i2=0;i2<n;i2++){
        b[i0][i1][i2]=0;
      }
    }
  }

#pragma xmp gmove
  b[0:n][0:n][0:n]=a[0:n][0:n][0:n];

  ierr=0;
#pragma xmp loop (i1,i2) on tx(*,i1,i2)
  for(i0=0;i0<n;i0++){
    for(i1=0;i1<n;i1++){
      for(i2=0;i2<n;i2++){
        ierr=ierr+abs(b[i0][i1][i2]-i0-i1-i2-1);
      }
    }
  }

#pragma xmp reduction (MAX:ierr)
  chk_int(name, ierr);

}

void gmove_bca_3a3t_b_s(){

char name[NAMELEN]="gmove_bca_3a3t_b_s";
int n=8;
int a[n][n][n],b[n][n][n];
#pragma xmp nodes p(2,2,2)
#pragma xmp nodes q(2,2,2)
#pragma xmp template tx(0:n-1,0:n-1,0:n-1)
#pragma xmp template ty(0:n-1,0:n-1,0:n-1)
#pragma xmp distribute tx(block,block,block) onto p
#pragma xmp distribute ty(block,block,block) onto q
#pragma xmp align a[i0][i1][i2] with tx(i0,i1,i2)
#pragma xmp align b[*][i1][i2] with ty(*,i1,i2)
#pragma xmp shadow a[0][0][1]
#pragma xmp shadow b[0][0][1]

  int i0,i1,i2,ierr;

#pragma xmp loop (i0,i1,i2) on tx(i0,i1,i2)
  for(i0=0;i0<n;i0++){
    for(i1=0;i1<n;i1++){
      for(i2=0;i2<n;i2++){
        a[i0][i1][i2]=i0+i1+i2+1;
      }
    }
  }

#pragma xmp loop (i1,i2) on ty(*,i1,i2)
  for(i0=0;i0<n;i0++){
    for(i1=0;i1<n;i1++){
      for(i2=0;i2<n;i2++){
        b[i0][i1][i2]=0;
      }
    }
  }

#pragma xmp gmove
  b[0:n][0:n][0:n]=a[0:n][0:n][0:n];

  ierr=0;
#pragma xmp loop (i1,i2) on ty(*,i1,i2)
  for(i0=0;i0<n;i0++){
    for(i1=0;i1<n;i1++){
      for(i2=0;i2<n;i2++){
        ierr=ierr+abs(b[i0][i1][i2]-i0-i1-i2-1);
      }
    }
  }

#pragma xmp reduction (MAX:ierr)
  chk_int(name, ierr);

}

void gmove_bca_3a4t_b(){

char name[NAMELEN]="gmove_bca_3a4t_b";
int n=8;
int a[n][n][n],b[n][n][n];
#pragma xmp nodes p(2,2,2)
#pragma xmp template tx(0:n-1,0:n-1,0:n-1,0:n-1)
#pragma xmp template ty(0:n-1,0:n-1,0:n-1,0:n-1)
#pragma xmp distribute tx(block,block,*,block) onto p
#pragma xmp distribute ty(block,*,block,block) onto p
#pragma xmp align a[*][i1][i3] with tx(*,i1,*,i3)
#pragma xmp align b[i0][i2][*] with ty(i0,*,i2,*)

  int i0,i1,i2,i3,ierr;

  for(i0=0;i0<n;i0++){
#pragma xmp loop (i1,i3) on tx(*,i1,*,i3)
    for(i1=0;i1<n;i1++){
      for(i3=0;i3<n;i3++){
        a[i0][i1][i3]=i0+i1+i3+1;
      }
    }
  }

#pragma xmp loop (i0,i2) on ty(i0,*,i2,*)
  for(i0=0;i0<n;i0++){
    for(i2=0;i2<n;i2++){
      for(i3=0;i3<n;i3++){
        b[i0][i2][i3]=0;
      }
    }
  }

#pragma xmp gmove
  b[0:n][0:n][0:n]=a[0:n][0:n][0:n];

  ierr=0;
#pragma xmp loop (i0,i2) on ty(i0,*,i2,*)
  for(i0=0;i0<n;i0++){
    for(i2=0;i2<n;i2++){
      for(i3=0;i3<n;i3++){
        ierr=ierr+abs(b[i0][i2][i3]-i0-i2-i3-1);
      }
    }
  }

#pragma xmp reduction (MAX:ierr)
  chk_int(name, ierr);

}

int main(){

  gmove_bca_2a3t_b2_subcomm();
  gmove_bca_2a3t_b();
  gmove_bca_3a3t_b2();
  gmove_bca_3a3t_b();
  gmove_bca_3a3t_b_s();
  gmove_bca_3a4t_b();

  return 0;

}

File: tests/XMP/global-view/gmove/F/gmove_bca_8.f90
program gmove_bca_8

  character*25 tname
  call gmove_bca_1a3t_b(tname)
  call gmove_bca_2a2t_b_subcomm(tname)
  call gmove_bca_2a3t_b(tname)
  call gmove_bca_3a3t_b2(tname)
  call gmove_bca_3a3t_b(tname)
  call gmove_bca_3a3t_bc2(tname)
  call gmove_bca_3a3t_bc(tname)
  call gmove_bca_3a4t_b(tname)

end program

subroutine gmove_bca_1a3t_b(tname)

character(*) tname
integer :: i
integer,parameter :: n=4
integer a(n),b(n)
!$xmp nodes p(2,2,2)
!$xmp template tx(n,n,n)
!$xmp distribute tx(block,block,block) onto p
!$xmp align a(i1) with tx(*,i1,*)
!$xmp align b(i2) with tx(*,*,i2)

!$xmp loop (i1) on tx(*,i1,*)
do i1=1,n
  a(i1)=i1
end do

!$xmp loop (i2) on tx(*,*,i2)
do i2=1,n
  b(i2)=0
end do

!$xmp gmove
b(1:n)=a(1:n)

myrank=xmp_node_num()
ierr=0
!$xmp loop (i2) on tx(*,*,i2)
do i2=1,n
  ierr=ierr+abs(b(i2)-i2)
end do

!$xmp reduction (+:ierr)
tname="gmove_bca_1a3t_b"
call chk_int(tname, ierr)

end subroutine

subroutine gmove_bca_2a2t_b_subcomm(tname)

character(*) tname
integer :: i
integer,parameter :: n=8
integer a(n,n),b(n,n)
!$xmp nodes p(8)
!$xmp nodes p1(2,2)=p(1:4)
!$xmp nodes p2(2,2)=p(5:8)
!$xmp template tx(n,n)
!$xmp template ty(n,n)
!$xmp distribute tx(block,block) onto p1
!$xmp distribute ty(block,block) onto p2
!$xmp align a(i0,i1) with tx(i0,i1)
!$xmp align b(*,i1) with ty(*,i1)

!$xmp loop (i0,i1) on tx(i0,i1)
do i1=1,n
  do i0=1,n
    a(i0,i1)=i0+i1
  end do
end do

!$xmp loop (i1) on ty(*,i1)
do i1=1,n
  do i0=1,n
    b(i0,i1)=0
  end do
end do

!$xmp gmove
b(1:n,1:n)=a(1:n,1:n)

ierr=0
!$xmp loop (i1) on ty(*,i1)
do i1=1,n
  do i0=1,n
    ierr=ierr+abs(b(i0,i1)-i0-i1)
  end do
end do

!$xmp reduction (+:ierr) on p2
tname="gmove_bca_2a2t_b_subcomm"
call chk_int(tname, ierr)

end subroutine

subroutine gmove_bca_2a3t_b(tname)

character(*) tname
integer,parameter :: n=4
integer a(n,n),b(n,n)
!$xmp nodes p(2,2,2)
!$xmp template tx(n,n,n)
!$xmp distribute tx(block,block,block) onto p
!$xmp align a(i0,i1) with tx(i0,i1,*)
!$xmp align b(i1,i2) with tx(*,i1,i2)

!$xmp loop (i0,i1) on tx(i0,i1,*)
do i1=1,n
  do i0=1,n
    a(i0,i1)=i0+i1
  end do
end do

!$xmp loop (i1,i2) on tx(*,i1,i2)
do i2=1,n
  do i1=1,n
    b(i1,i2)=0
  end do
end do

!$xmp gmove
b(:,:)=a(:,:)

myrank=xmp_node_num()
ierr=0
!$xmp loop (i1,i2) on tx(*,i1,i2)
do i2=1,n
  do i1=1,n
    ierr=ierr+abs(b(i1,i2)-i1-i2)
  end do
end do

!$xmp reduction (+:ierr)
tname="gmove_bca_2a3t_b"
call chk_int(tname, ierr)

end subroutine

subroutine gmove_bca_3a3t_b2(tname)

character(*) tname
integer :: i
integer,parameter :: n=4
integer a(n,n,n),b(n,n,n)
!$xmp nodes p(2,2,2)
!$xmp template tx(n,n,n)
!$xmp distribute tx(block,block,block) onto p
!$xmp align a(i0,i1,i2) with tx(i0,i1,i2)
!$xmp align b(*,i1,i2) with tx(*,i1,i2)

!$xmp loop (i0,i1,i2) on tx(i0,i1,i2)
do i2=1,n
  do i1=1,n
    do i0=1,n
      a(i0,i1,i2)=i0+i1+i2
    end do
  end do
end do

!$xmp loop (i1,i2) on tx(*,i1,i2)
do i2=1,n
  do i1=1,n
    do i0=1,n
      b(i0,i1,i2)=0
    end do
  end do
end do

!$xmp gmove
b(:,:,:)=a(:,:,:)

ierr=0
!$xmp loop (i1,i2) on tx(*,i1,i2)
do i2=1,n
  do i1=1,n
    do i0=1,n
      ierr=ierr+abs(b(i0,i1,i2)-i0-i1-i2)
    end do
  end do
end do

!$xmp reduction (+:ierr)
tname="gmove_bca_3a3t_b2"
call chk_int(tname, ierr)

end subroutine

subroutine gmove_bca_3a3t_b(tname)

character(*) tname
integer :: i
integer,parameter :: n=4
integer a(n,n,n),b(n,n,n)
!$xmp nodes p(2,2,2)
!$xmp template tx(n,n,n)
!$xmp distribute tx(block,block,block) onto p
!$xmp align a(i0,i1,i2) with tx(i0,i1,i2)
!$xmp align b(*,i1,i2) with tx(*,i1,i2)

!$xmp loop (i0,i1,i2) on tx(i0,i1,i2)
do i2=1,n
  do i1=1,n
    do i0=1,n
      a(i0,i1,i2)=i0+i1+i2
    end do
  end do
end do

!$xmp loop (i1,i2) on tx(*,i1,i2)
do i2=1,n
  do i1=1,n
    do i0=1,n
      b(i0,i1,i2)=0
    end do
  end do
end do

!$xmp gmove
b(1:n,1:n,1:n)=a(1:n,1:n,1:n)

ierr=0
!$xmp loop (i1,i2) on tx(*,i1,i2)
do i2=1,n
  do i1=1,n
    do i0=1,n
      ierr=ierr+abs(b(i0,i1,i2)-i0-i1-i2)
    end do
  end do
end do

!$xmp reduction (+:ierr)
tname="gmove_bca_3a3t_b"
call chk_int(tname, ierr)

end subroutine

subroutine gmove_bca_3a3t_bc2(tname)

character(*) tname
integer :: i
integer,parameter :: n=4
integer a(n,n,n),b(n,n,n)
!$xmp nodes p(2,2,2)
!$xmp template tx(n,n,n)
!$xmp distribute tx(cyclic(2),cyclic(2),cyclic(2)) onto p
!$xmp align a(i0,i1,i2) with tx(i0,i1,i2)
!$xmp align b(*,i1,i2) with tx(*,i1,i2)

!$xmp loop (i0,i1,i2) on tx(i0,i1,i2)
do i2=1,n
  do i1=1,n
    do i0=1,n
      a(i0,i1,i2)=i0+i1+i2
    end do
  end do
end do

!$xmp loop (i1,i2) on tx(*,i1,i2)
do i2=1,n
  do i1=1,n
    do i0=1,n
      b(i0,i1,i2)=0
    end do
  end do
end do

!$xmp gmove
b(:,:,:)=a(:,:,:)

ierr=0
!$xmp loop (i1,i2) on tx(*,i1,i2)
do i2=1,n
  do i1=1,n
    do i0=1,n
      ierr=ierr+abs(b(i0,i1,i2)-i0-i1-i2)
    end do
  end do
end do

!$xmp reduction (+:ierr)
tname="gmove_bca_3a3t_bc2"
call chk_int(tname, ierr)

end subroutine

subroutine gmove_bca_3a3t_bc(tname)

character(*) tname
integer :: i
integer,parameter :: n=4
integer a(n,n,n),b(n,n,n)
!$xmp nodes p(2,2,2)
!$xmp template tx(n,n,n)
!$xmp distribute tx(cyclic(2),cyclic(2),cyclic(2)) onto p
!$xmp align a(i0,i1,i2) with tx(i0,i1,i2)
!$xmp align b(*,i1,i2) with tx(*,i1,i2)

!$xmp loop (i0,i1,i2) on tx(i0,i1,i2)
do i2=1,n
  do i1=1,n
    do i0=1,n
      a(i0,i1,i2)=i0+i1+i2
    end do
  end do
end do

!$xmp loop (i1,i2) on tx(*,i1,i2)
do i2=1,n
  do i1=1,n
    do i0=1,n
      b(i0,i1,i2)=0
    end do
  end do
end do

!$xmp gmove
b(2:n,2:n,2:n)=a(2:n,2:n,2:n)

ierr=0
!$xmp loop (i1,i2) on tx(*,i1,i2)
do i2=2,n
  do i1=2,n
    do i0=2,n
      ierr=ierr+abs(b(i0,i1,i2)-i0-i1-i2)
    end do
  end do
end do

!$xmp reduction (+:ierr)
tname="gmove_bca_3a3t_bc"
call chk_int(tname, ierr)

end subroutine

subroutine gmove_bca_3a4t_b(tname)

character(*) tname
integer :: i
integer,parameter :: n=4
integer a(n,n,n),b(n,n,n)
!$xmp nodes p(2,2,2)
!$xmp template tx(n,n,n,n)
!$xmp template ty(n,n,n,n)
!$xmp distribute tx(block,block,*,block) onto p
!$xmp distribute ty(block,*,block,block) onto p
!$xmp align a(*,i1,i3) with tx(*,i1,*,i3)
!$xmp align b(i0,i2,*) with ty(i0,*,i2,*)

!$xmp loop (i1,i3) on tx(*,i1,*,i3)
do i3=1,n
  do i1=1,n
    do i0=1,n
      a(i0,i1,i3)=i0+i1+i3
    end do
  end do
end do

do i3=1,n
!$xmp loop (i0,i2) on ty(i0,*,i2,*)
  do i2=1,n
    do i0=1,n
      b(i0,i1,i2)=0
    end do
  end do
end do

!$xmp gmove
b(1:n,1:n,1:n)=a(1:n,1:n,1:n)

ierr=0
do i3=1,n
!$xmp loop (i0,i2) on ty(i0,*,i2,*)
  do i2=1,n
    do i0=1,n
      ierr=ierr+abs(b(i0,i2,i3)-i0-i2-i3)
    end do
  end do
end do

!$xmp reduction (+:ierr)
tname="gmove_bca_3a4t_b"
call chk_int(tname, ierr)

end subroutine
